### DISCUSSION TITLE
测试版`1.9.2-beta`专项讨论

### DISCUSSION THREAD

**Discussion by maboloshi** [maboloshi/github-chinese#146] [General]
> [!TIP]
  > 测试脚本地址：
[zh-CN_LangEnvTest](https://github.com/maboloshi/github-chinese/tree/zh-CN_LangEnvTest) 分支下的 [main.user.js](https://github.com/maboloshi/github-chinese/raw/zh-CN_LangEnvTest/main.user.js)

- [x] 进一步优化中文环境设置
   - [x] 当前使用`document-start`脚本载入机制, 导致存在一些 [bug](146#discussioncomment-9696969), 打算使用`window.addEventListener('DOMContentLoaded', init);`延迟`init`的载入 (灵感来自 buiawpkgew1/github-chinese#98)
- [x] 使用 ~~`window.addEventListener('DOMContentLoaded', ...)`~~ `document.addEventListener('turbo:load', ...)`代替`setTimeout`
- [ ] ~~更新`page`更新逻辑~~
   - [ ] ~~使用`Turbo`原生的`before-fetch-request`事件及`detail.url`对象 最快获取新页面的`url`. 目前,发现对于需要`class`筛选的效果不好~~
   - [ ] ~~监视`document.body`的`class`属性变化即可~~
- [ ] 词库清理计划 (#115)
- [x] 翻译 GitHub 一些附属网站，如 `githubstatus.com` 与 `skills.github.com`
- [x] 修复 https://github.com/maboloshi/github-chinese/discussions/146#discussioncomment-9855443 ，并优化`getPage`函数
- [x] 优化`traverseNode`函数
- [x] 优化`transElement`函数
- [ ] 优化`watchUpdate`函数:
  - [x] 突变处理
     - [x] 在`mutations.flatMap`过程：
        - [x] 对于`节点增加`突变，后期迭代翻译的对象调整为`addedNodes`中记录的新增节点（大幅减少重复迭代的问题），其他突变处理不变
        - [x] 丢弃掉特定页面的特定元素，包括不限于修复代码新建和编辑页面触发逗号 bug 的问题（配合`ignoreSelector`和`reIgnoreClass`等规则）；
        - [x] 仅保留特定页面的`字符数据`的突变
        - [x] 重新生成一个包含突变节点的数组，用于迭代翻译
        - [ ] ~~进一步探讨`MutationObserver.takeRecords()` (灵感来自 buiawpkgew1/github-chinese#98)~~
   - [x] 将`监测 HTML lang 值`和`Turbo 事件监听器`剥离出来，使`watchUpdate`函数专注于监视页面的 DOM 变化

**Comment by maboloshi** [comment]
目前, 时间不会处理`on`
![image](https://github.com/maboloshi/github-chinese/assets/7850715/b9b208ac-b1ed-4554-bf56-8f7313c7d31c)

**Reply by maboloshi** [reply]
> 目前, 时间不会处理`on` ![image](https://private-user-images.githubusercontent.com/7850715/334823938-b9b208ac-b1ed-4554-bf56-8f7313c7d31c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc3MjY0MjYsIm5iZiI6MTcxNzcyNjEyNiwicGF0aCI6Ii83ODUwNzE1LzMzNDgyMzkzOC1iOWIyMDhhYy1iMWVkLTQ1NTQtYmY1Ni04ZjczMTNjN2QzMWMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MDdUMDIwODQ2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzgwMWNmYzY0N2E3MmJkOGEyOTdhNmI2NmU0N2VmODc4NWYyOWQ5ZDIxNTkwZjZlMmQzNWU0MjE1OGIyODZjMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.KqArCa9_wF2AO2opj7r-wyEDoeZRxXXt99Wnx7iz_9k)

已修复(c1a2ffe) 有问题请反馈

**Comment by maboloshi** [comment]
好像不一定每次都能设置`html`属性`lang`成功 抓狂
![Suggestion](https://github.com/maboloshi/github-chinese/assets/7850715/b49f9d8a-be1b-4eb3-8550-6a878c535391)

**Comment by maboloshi** [comment]
我有点傻了, 可以跟踪 `html`元素属性`lang`值的变化, 重新设置`lang`

**Comment by cosing** [comment]
不会js什么的，但是我有补充的信息

```js
  var times = document.getElementsByTagName('relative-time');
  for (var i = 0; i < times.length; i++) {
    times[i].update()
    //console.log(i)
  }
```
加了这段后运行时机在何时都无所谓了，但是似乎从库进入commit的时候会重新设置lang，并且此时也不会重新运行脚本，然后我问gpt要了个观察者什么的，阻止了这种改变。

我认为使用这种方法的好处是可以避免替换带来的某些错误，坏处是更依赖于GitHub本身的表现。

**Comment by maboloshi** [comment]
目前 测试其实并不充分

**Comment by maboloshi** [comment]
存在一定概率, `body`未载入, 无法判定是否存在`logged-in`
![image](https://github.com/maboloshi/github-chinese/assets/7850715/133d6010-7f75-4a32-98bf-0c3448b0d2ba)

**Reply by maboloshi** [reply]
已修复

**Comment by maboloshi** [comment]
@wuyuncheng-26 @TC999 @cosing 大家方便测试一下 [Test_zh-CN_LangEnvSet](https://github.com/maboloshi/github-chinese/tree/Test_zh-CN_LangEnvSet) 分支下的 [main.user.js](https://github.com/maboloshi/github-chinese/raw/Test_zh-CN_LangEnvSet/main.user.js) 么😊

1. 使用`document.documentElement.lang = 'zh-CN'`设置中文环境
2. 使用`document-start`脚本载入机制
3. 使用`window.addEventListener('DOMContentLoaded', init);`延迟`init()`的载入
4. 通过调整时间元素翻译和时间元素正则, 修复时间`on`的问题
5. 删除多余`watchTimeElement 函数`
6. 使用`document.addEventListener('turbo:load', ...)`代替`setTimeout`
7. 初始化部分代码

请大家进行广泛测试, 发现和修补 bug, 等待稳定后再合并到主分支

**Comment by TC999** [comment]
现在有概率不翻译时间，刷新一下又可以了

**Reply by maboloshi** [reply]
最好能截图个看看

**Reply by TC999** [reply]
> 最好能截图个看看

已经修了

**Comment by TC999** [comment]
等下，BUG出现了，在GitHub 操作页面，这次录到了影片 
示例页面<https://github.com/maboloshi/github-chinese/actions/runs/9436880727>

https://github.com/maboloshi/github-chinese/assets/88823709/122f7e55-6737-40d9-902c-249d531be7af

**Reply by maboloshi** [reply]
看着不像使用了测试脚本
最新的时间格式是浏览器默认(大陆简中)的
![image](https://github.com/maboloshi/github-chinese/assets/7850715/c4bf0b7e-e96c-47a3-adc1-3bb95ea1d534)

**Reply by maboloshi** [reply]
> 我觉得还是不要用 `document.documentElement.lang = 'zh';` 了吧 副作用有那么一点点多（数字和单位间空格消失了） 不如直接正则翻译

减少时间正则和时间元素的监视 理论上性能提升

**Reply by maboloshi** [reply]
> Action 可能没办法了，将错就错吧 `document-start` 本来就挺多问题的 我试过，用这个有时候可能有一部分内容直接不翻译了

这个应该是初始化函数提前运行的原因, 现在使用`window.addEventListener('DOMContentLoaded', init);`延迟运行, 理论上是修复了

**Reply by cosing** [reply]
document-start 倒是可解，这个只更新一次页面全部元素就不需要了
```js
var times = document.getElementsByTagName('relative-time');
for (var i = 0; i < times.length; i++) {
  times[i].update()
}
```

**Reply by maboloshi** [reply]
> 嗯 不过这个空格问题是不打算管了吗（ orz

是的, 这个你可以看一下 其他中文网站 基本就是这种格式

**Reply by maboloshi** [reply]
> document-start 倒是可解，这个只更新一次页面全部元素就不需要了
> 
> ```js
> var times = document.getElementsByTagName('relative-time');
> for (var i = 0; i < times.length; i++) {
>   times[i].update()
> }
> ```

目前, 已经没必要了

**Reply by maboloshi** [reply]
嗯 是的, 性能优先. 另外, 仅仅是空格问题看习惯也就这样了

**Reply by TC999** [reply]
其实没有多大影响，F5重载页面又正常了，并且直到关闭浏览器都不会再出现

**Comment by maboloshi** [comment]
`on`的问题貌似没有完全修复, 已经是唯一的一条时间元素正则了

**Comment by maboloshi** [comment]
接一下来是`watchUpdate`和`getPage`函数优化了, 基本脱离这个讨论议题了

**Comment by maboloshi** [comment]
另外, 后期统一一下测试版的版本号构建方式了. 实现, 测试版的脚本管理器自动更新

**Comment by TC999** [comment]
建议：把`Test_zh-CN_LangEnvSet`分支中的`README.md`中的第300行改为
```
[main.user.js]: https://github.com/maboloshi/github-chinese/raw/Test_zh-CN_LangEnvSet/main.user.js "GitHub 中文化插件 - GitHub 托管"
```
这样可以点击链接直接装

**Reply by maboloshi** [reply]
测试还是在小范围内测试比较好

**Comment by TC999** [comment]
## 最新情况
居然全是英文没翻译
将词条文件换成`gh-pages`分支也没用
- 如图所示
![image](https://github.com/maboloshi/github-chinese/assets/88823709/cc979712-5ab2-4cbc-b09a-b34abe63bacf)

**Reply by maboloshi** [reply]
修复了词库错误，请手动更新重试

**Reply by TC999** [reply]
> 修复了词库错误，请手动更新重试

修好了

**Comment by maboloshi** [comment]
目前，不管是正式版还是测试版均存在以下 bug：
比如：在仓库相关页面，从左上角点击“我的资料” 调转到“个人资料页” 会存在大概率的翻译不完全

大致原因是`body`的`class`更新较晚，导致不能第一时间获得到准确的`page`。后期考虑除了`session-authentication`这`class`规则保留，`page-profile`使用`pathname`规则代替。（ps：`page-new-repo`，`page-create-org`早已失效）

**Reply by maboloshi** [reply]
已尝试修复 c2dd05d

**Comment by maboloshi** [comment]
@wuyuncheng-26 @TC999 测试分支已更新，并更新脚本名称，便于脚本管理器中同时安装2种版本便于调试
请大家删除已安装的老版本测试分支脚本，重新安装。安装地址不变

**Reply by maboloshi** [reply]
再次重大更新
![03C3F030](https://github.com/maboloshi/github-chinese/assets/7850715/88fbaca2-fcac-4b97-b64a-13dc6e6ff0dd)

**Comment by maboloshi** [comment]
主分支 1.9.3 已发布, 等公测没大问题就推送`greasyfork版本`
